File: 백준/Gold/17298. 오큰수/오큰수.java
import java.util.*;

public class Main {
    public static void main(String[] args) {
        Stack<Integer> st = new Stack<>();
        Scanner sc = new Scanner(System.in);

        int n = sc.nextInt();
        int[] arr = new int[n];
        int[] answer = new int[n];

        for (int i = 0; i < n; i++) {
            arr[i] = sc.nextInt();
        }
        for(int i = 0; i < n; i++){
            while(!st.isEmpty() && arr[st.peek()] < arr[i]){
                answer[st.pop()] = arr[i];
            }
            st.push(i);
        }
        while(!st.isEmpty()){
            answer[st.pop()] = -1;
        }
        StringBuilder sb = new StringBuilder();
        for(int i = 0; i <n; i++){
            sb.append(answer[i]).append(" ");
        }
        System.out.println(sb);
    }
}

File: 백준/Gold/17298. 오큰수/오큰수.kt
import java.util.*

fun main(){
    val sc = Scanner(System.`in`)
    val n = sc.nextInt()
    var num = IntArray(n)
    var result = IntArray(n)
    val st = Stack<Int>()

    for(i in 0 until n){
        num[i] = sc.nextInt()
    }

    for(i in 0 until n){
        while(st.isNotEmpty() && num[st.peek()] < num[i]){
            result[st.pop()] = num[i]
        }
        st.push(i)
    }
    while(st.isNotEmpty()){
        result[st.pop()] = -1
    }
    val sb = StringBuilder()
    for(n in result){
        sb.append(n).append(" ")
    }
    println(sb)
}